cmd/kops/rotate_ca.go

```diff
@@ -0,0 +1,310 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/cmd/kops/util"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/apis/kops/model"
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+var (
+	rotateCALong = templates.LongDesc(i18n.T(`
+	rotates the cluster CA.`))
+
+	rotateCAExample = templates.Examples(i18n.T(`
+	# Rotate the cluster CA
+	kops rotate ca --name k8s-cluster.example.com
+	`))
+
+	rotateCAShort = i18n.T(`Rotates the cluster CA.`)
+)
+
+func NewCmdRotateCA(f *util.Factory, out io.Writer) *cobra.Command {
+
+	cmd := &cobra.Command{
+		Use:     "ca",
+		Long:    rotateCALong,
+		Short:   rotateCAShort,
+		Example: rotateCAExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.Background()
+
+			err := rootCommand.ProcessArgs(args)
+			if err != nil {
+				exitWithError(err)
+			}
+
+			clusterName := rootCommand.ClusterName()
+
+			if err := RunRotateCA(ctx, f, clusterName, out); err != nil {
+				exitWithError(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func RunRotateCA(ctx context.Context, f *util.Factory, clusterName string, out io.Writer) error {
+
+	cluster, err := rootCommand.Cluster(ctx)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(os.Stdin)
+
+	contextName := cluster.ObjectMeta.Name
+	clientConfig, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+
+	if err != nil {
+		return fmt.Errorf("failed to delete secondary key: %w", err)
+	}
+
+	if !model.UseKopsControllerForNodeBootstrap(cluster) {
+		return fmt.Errorf("only clusters using kops-controller for boostrapping nodes are supported")
+	}
+
+	exportAdmin := clientConfig.Contexts[contextName].AuthInfo == contextName
```

Comment:
This appears to be a bit too heuristic. It should check that the corresponding `AuthInfo` has nonempty `ClientCertificateData`